### PR TITLE
Proof-of-stake-time

### PR DESCRIPTION
**Pull request by SkidanovAlex**
- Replace the weight of a block with "proof of stake time" -- the percentage of the total stake multiplied by reported time between the two previous blocks.
- Add missing validation of the score of blocks.
- Pro-rating the weight if a block is in the future compared to our current timestamp

Test plan:
- Added a new test that verifies:
   - that a longer chain (in terms of number of blocks) with fewer approvals per block cannot win given the same timestamps
   - that a block that is almost two minutes into the future won't get accepted as the head
   - that a block that is almost two minutes into the future with > 1/2 approvals does become a head
   - also creates a block in near future and confirms that it doesn't become head early but does become if something is built upon it when the "close future" is already past

Also had to change the time computation for blocks in the test infra to use 20ns increments, to simplify computing the weight in the tests that do not target the fork choice rule.

### CODE REVIEW

File: chain/chain/src/types.rs [hunk 1]
```diff
@@ -444,6 +441,38 @@ pub trait RuntimeAdapter: Send + Sync {
...
+            approved_stake += *account_to_stake.get(approval).unwrap_or(&0u128);
+            account_to_stake.insert(approval.clone(), 0);
+        }
+
+        let ret = if total_stake >= std::u64::MAX as u128 {
+            approved_stake / (total_stake / WEIGHT_MULTIPLIER)
+        } else {
+            WEIGHT_MULTIPLIER * approved_stake / total_stake
+        };
+        return Ok(max(ret, 1));
```
Comment (bowenwang1996) [inline chain/chain/src/types.rs]: you don't need `return`

File: chain/chain/tests/simple_chain.rs [hunk 2]
```diff
@@ -19,6 +19,7 @@ fn build_chain() {
         let prev_hash = chain.head_header().unwrap().hash();
         let prev = chain.get_block(&prev_hash).unwrap();
         let block = Block::empty(&prev, &*signer);
+        println!("{}", block.header.inner_rest.total_weight);
```
Comment (bowenwang1996) [inline chain/chain/tests/simple_chain.rs]: remove?

File: core/primitives/src/block.rs [hunk 3]
```diff
@@ -477,11 +479,12 @@ impl Block {
         let new_total_supply =
             prev.inner_rest.total_supply + inflation.unwrap_or(0) - balance_burnt;
 
-        let num_approvals: u128 = approvals.len() as u128;
-        let total_weight = prev.inner_rest.total_weight.next(num_approvals);
         let now = to_timestamp(Utc::now());
         let time =
             if now <= prev.inner_lite.timestamp { prev.inner_lite.timestamp + 1 } else { now };
+        let time_delta = if time_delta < 1 { 1 } else { time_delta };
```
Comment (evgenykuzyakov) [inline core/primitives/src/block.rs]: by `< 1` you mean zero? sounds like `max(1, time_delta)` might work as well

File: chain/chain/src/types.rs [hunk 4]
```diff
@@ -444,6 +441,38 @@ pub trait RuntimeAdapter: Send + Sync {
...
+            account_to_stake.insert(bp.0.account_id, bp.0.amount);
+            total_stake += bp.0.amount;
+        }
+
+        for approval in approvals {
+            approved_stake += *account_to_stake.get(approval).unwrap_or(&0u128);
+            account_to_stake.insert(approval.clone(), 0);
+        }
+
+        let ret = if total_stake >= std::u64::MAX as u128 {
```
[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply chain/chain/src/types.rs]: The code here is just to handle overflows.
`u64::MAX` should be `u128::MAX / WEIGHT_MULTIPLIER`, replacing.

**Pull request merged**